streaming/base/batching/stratified.py

```diff
@@ -0,0 +1,171 @@
+                f'of this stream in the global batch, which is of size {global_batch_size}, is ' +
+                f'too low. Please increase the global batch size or increase the porportion of ' +
+                f'total samples that come from stream {stream_id}.')
+
+    # We now merge the partitions from each stream to get our final partition over all streams, where
+    # every single global batch has the same sample composition from the streams.
+    # The total number of batches we can make is constrained by the min batch parts available from any one stream.
+    min_batch_parts = np.inf
+    batches_from_partitions = []
+    for i, partition in enumerate(partition_per_stream):
+        # Reshape the partition to batch portion per stream in order of traversal, and count only batches without -1 in them.
+        # Before reshaping, make sure number of samples in each stream is divisible by the batch_portion_per_stream
+        batch_parts_inorder = partition.transpose(3, 2, 0, 1, 4).flatten()
+        samples_in_stream_partition = batch_parts_inorder.size
+        if samples_in_stream_partition % batch_portion_per_stream[i] != 0:
+            padding_samples = batch_portion_per_stream[i] - (samples_in_stream_partition %
+                                                             batch_portion_per_stream[i])
+            batch_parts_inorder = np.concatenate(
+                (batch_parts_inorder, np.full(padding_samples, -1)))
+        # Reshape to get batch portions from this stream, in order of traversal.
+        batch_parts_inorder = batch_parts_inorder.reshape(-1, batch_portion_per_stream[i])
+        num_full_batches = np.count_nonzero(np.min(batch_parts_inorder, axis=1) >= 0)
+        if num_full_batches != batch_parts_inorder.shape[0]:
+            logger.warning(
+                'Because of the `stratified` batching method, some batches with an inadequate number of samples '
+                + 'from stream with index ' + str(i) + ' are being dropped.')
+        if num_full_batches < min_batch_parts:
+            min_batch_parts = num_full_batches
+        batches_from_partitions.append(batch_parts_inorder)
+
+    # clip the partitions from all streams to only have min_batch_parts batch parts.
+    batches_from_partitions = [
+        batch_partition[:min_batch_parts] for batch_partition in batches_from_partitions
+    ]
+
+    # Concatenate the batch parts from every stream to form all the global batches
+    all_partition_batches = np.concatenate(batches_from_partitions, axis=1)
+
+    # If applicable we resume right after the most recently used full global batch.
+    if sample_in_epoch % global_batch_size != 0:
+        logger.warning(
+            'Because of the `stratified` batching method, resumption may only occur on a sample that '
+            + 'is a multiple of the current global batch size of ' + str(global_batch_size) +
+            '. Resuming training ' + 'after the most recently finished global batch.')
```

Comment:
Do you need `+` here ?

When would this warning occurs ? During resumption, if global_batch_size changes, then they should see this message right? if that's the case, should we change this `warnings.warn` ?

Reply:
This error message would happen if they tried to resume on a different (non-divisible) global batch size or if somehow `sample_in_epoch` was in the middle of a batch.